Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/detect_features.h

```diff
@@ -152,77 +165,72 @@ detect_surface_patches(PolygonMesh& p,
 
 template <typename PolygonMesh, typename EdgeIsFeatureMap, typename PatchIdMap>
 typename boost::graph_traits<PolygonMesh>::faces_size_type
-detect_surface_patches(PolygonMesh& p,
+detect_surface_patches(const PolygonMesh& p,
                        PatchIdMap patch_id_map,
                        EdgeIsFeatureMap eif)
 {
   return detect_surface_patches(p, patch_id_map, eif, parameters::all_default());
 }
 
 
-template<typename GT,
-         typename FT,
-         typename PolygonMesh,
-         typename EIFMap,
-         typename VNFEMap>
- void sharp_call(PolygonMesh& pmesh,
-                 FT angle_in_deg,
-                 EIFMap edge_is_feature_map,
-                 VNFEMap vnfe)
+template <typename FT, typename PolygonMesh, typename VPM, typename GT, typename EIFMap, typename VNFEMap>
+void sharp_call(const FT angle_in_deg,
+                const PolygonMesh& pmesh,
+                const VPM vpm,
+                const GT gt,
+                EIFMap edge_is_feature_map,
+                VNFEMap vnfe)
 {
   // Initialize vertices
-  for(typename boost::graph_traits<PolygonMesh>::vertex_descriptor vd :
-                vertices(pmesh))
-  {
+  for(typename boost::graph_traits<PolygonMesh>::vertex_descriptor vd : vertices(pmesh))
     put(vnfe, vd, 0);
-  }
-  FT cos_angle ( std::cos(CGAL::to_double(angle_in_deg) * CGAL_PI / 180.) );
+
+  const FT cos_angle = std::cos(CGAL::to_double(angle_in_deg) * CGAL_PI / 180.);
+  const FT sq_cos_angle = square(cos_angle);
 
   // Detect sharp edges
   for(typename boost::graph_traits<PolygonMesh>::edge_descriptor ed : edges(pmesh))
   {
-    typename boost::graph_traits<PolygonMesh>::halfedge_descriptor he = halfedge(ed,pmesh);
-    if(is_border_edge(he,pmesh)
-      || angle_in_deg == FT()
-      || (angle_in_deg != FT(180) && internal::is_sharp<PolygonMesh, GT>(pmesh,he,cos_angle))
-      )
+    typename boost::graph_traits<PolygonMesh>::halfedge_descriptor he = halfedge(ed, pmesh);
+    if(is_border_edge(he, pmesh) ||
+       angle_in_deg == FT() ||
+       (angle_in_deg != FT(180) && internal::is_sharp(he, pmesh, vpm, gt, sign(cos_angle), sq_cos_angle)))
```

Comment:
It should be `CGAL::sign`, to avoid the argument-dependent lookup.
```suggestion
       (angle_in_deg != FT(180) && internal::is_sharp(he, pmesh, vpm, gt, CGAL::sign(cos_angle), sq_cos_angle)))
```